### PR TITLE
feat: add improved search

### PR DESCRIPTION
**Pull request by Meierschlumpf**
<br/>
<div align="center">
  
  <h3>Homarr</h3>
</div>

**Thank you for your contribution. Please ensure that your pull request meets the following pull request:**

- [x] Builds without warnings or errors (``pnpm buid``, autofix with ``pnpm format:fix``)
- [x] Pull request targets ``dev`` branch
- [x] Commits follow the [conventional commits guideline](https://www.conventionalcommits.org/en/v1.0.0/)
- [x] No shorthand variable names are used (eg. ``x``, ``y``, ``i`` or any abbrevation)

### Some screenshots

image
image
image
image
image
image
image
image
image
image
image
image
image

### CODE REVIEW

File: packages/api/src/router/board.ts [hunk 1]
```diff
@@ -109,6 +109,20 @@ export const boardRouter = createTRPCRouter({
       isHome: currentUserWhenPresent?.homeBoardId === board.id,
     }));
   }),
+  search: publicProcedure
+    .input(z.object({ query: z.string(), limit: z.number().min(1).max(100).default(10) }))
+    .query(({ ctx, input }) => {
+      // TODO: restrict for access
```
Comment (Meierschlumpf) [inline packages/api/src/router/board.ts:115]: TODO

File: packages/spotlight/src/components/actions/group-actions.tsx [hunk 2]
```diff
@@ -0,0 +1,73 @@
...
+      .sort((optionA, optionB) => {
+        if ("sort" in group) {
+          return group.sort?.(query, [optionA, optionB]) ?? 0;
+        }
+
+        return 0;
+      });
+
+    if (filteredOptions.length === 0) {
+      return <Spotlight.Empty>{t("common.search.nothingFound")}</Spotlight.Empty>;
```
Comment (Meierschlumpf) [inline packages/spotlight/src/components/actions/group-actions.tsx:40]: Add shared component

File: packages/spotlight/src/components/actions/group-actions.tsx [hunk 3]
```diff
@@ -0,0 +1,73 @@
...
+        group={group}
+        query={query}
+        setMode={setMode}
+        setChildrenOptions={setChildrenOptions}
+      />
+    ));
+  }
+
+  if (options.isLoading || options.isError || !options.data) {
+    return <></>;
```
Comment (Meierschlumpf) [inline packages/spotlight/src/components/actions/group-actions.tsx:56]: Add loading and error components

File: packages/spotlight/src/components/actions/items/children-action-item.tsx [hunk 4]
```diff
@@ -0,0 +1,34 @@
...
+  query: string;
+  action: ReturnType<inferSearchInteractionOptions<"children">["useActions"]>[number];
+}
+
+export const ChildrenActionItem = ({ childrenOptions, action, query }: ChildrenActionItemProps) => {
+  const interaction = action.useInteraction(childrenOptions.option, query);
+
+  if (interaction.type === "disabled") {
+    return <></>;
+  }
```
Comment (Meierschlumpf) [inline packages/spotlight/src/components/actions/items/children-action-item.tsx:18]: Probably remove as there is hide & filter

File: packages/spotlight/src/components/spotlight.tsx [hunk 5]
```diff
@@ -0,0 +1,136 @@
...
+import type { inferSearchInteractionOptions } from "../lib/interaction";
+import { searchModes } from "../modes";
+import { searchEnginesChildrenOptions, searchEnginesSearchGroups } from "../modes/external/search-engines-search-group";
+import { selectAction, spotlightStore } from "../spotlight-store";
+import { SpotlightChildrenActions } from "./actions/children-actions";
+import { SpotlightActionGroups } from "./actions/groups/action-group";
+
+export const NewSpotlight = () => {
+  const [query, setQuery] = useState("");
+  const [mode, setMode] = useState("default");
```
Comment (Meierschlumpf) [inline packages/spotlight/src/components/spotlight.tsx:19]: Maybe change to `help` for now and change later once we have search engines and something to show on the start page

File: packages/spotlight/src/components/spotlight.tsx [hunk 6]
```diff
@@ -0,0 +1,136 @@
...
+          )
+        }
+        value={query}
+        onKeyDown={(event) => {
+          if (query.length === 0 && mode !== "default" && event.key === "Backspace") {
+            setMode("default");
+            setChildrenOptions(null);
+          }
+
+          // TODO: Add api to directly interact / maybe even add an option to add a onKeyPress event to modes / groups?
```
Comment (Meierschlumpf) [inline packages/spotlight/src/components/spotlight.tsx:82]: TODO

File: packages/spotlight/src/components/spotlight.tsx [hunk 7]
```diff
@@ -0,0 +1,136 @@
...
+        <Group>
+          <childrenOptions.detailComponent options={childrenOptions.option as never} />
+        </Group>
+      ) : (
+        <Group justify="start" px="md" py="sm">
+          {activeMode?.tip ? (
+            activeMode.tip
+          ) : (
+            <Text size="xs" c="gray.6">
+              Type <Kbd size="xs">?</Kbd> for help and tips
```
Comment (Meierschlumpf) [inline packages/spotlight/src/components/spotlight.tsx:107]: Add translation

File: packages/spotlight/src/components/spotlight.tsx [hunk 8]
```diff
@@ -0,0 +1,136 @@
...
+      )}
+
+      <Spotlight.ActionsList>
+        {childrenOptions ? (
+          <SpotlightChildrenActions childrenOptions={childrenOptions} query={query} />
+        ) : activeMode ? (
+          <SpotlightActionGroups
+            setMode={(mode) => {
+              setMode(mode);
+              setChildrenOptions(null);
```
Comment (Meierschlumpf) [inline packages/spotlight/src/components/spotlight.tsx:120]: Also set selectAction to 0

File: packages/spotlight/src/lib/group.ts [hunk 9]
```diff
@@ -0,0 +1,26 @@
+import type { UseTRPCQueryResult } from "@trpc/react-query/shared";
+
+import type { inferSearchInteractionDefinition, SearchInteraction } from "./interaction";
+
+type CommonSearchGroup<TOption extends Record<string, unknown>, TOptionProps extends Record<string, unknown>> = {
+  // key path is used to define the path to a unique key in the option object
+  keyPath: keyof TOption;
+  title: string;
```
Comment (Meierschlumpf) [inline packages/spotlight/src/lib/group.ts:8]: Make translatable

File: packages/spotlight/src/lib/mode.ts [hunk 10]
```diff
@@ -0,0 +1,11 @@
+import type { ReactNode } from "react";
+
+import type { SearchGroup } from "./group";
+
+export interface SearchMode {
+  name: string;
+  character: string;
+  help: string | undefined;
```
Comment (Meierschlumpf) [inline packages/spotlight/src/lib/mode.ts:8]: Make translatable

File: packages/spotlight/src/lib/mode.ts [hunk 11]
```diff
@@ -0,0 +1,11 @@
+import type { ReactNode } from "react";
+
+import type { SearchGroup } from "./group";
+
+export interface SearchMode {
+  name: string;
```
Comment (Meierschlumpf) [inline packages/spotlight/src/lib/mode.ts:6]: Make translatable

File: packages/spotlight/src/modes/app-integration-board/apps-search-group.tsx [hunk 12]
```diff
@@ -0,0 +1,80 @@
...
+type App = { id: string; name: string; iconUrl: string; href: string | null };
+
+const appChildrenOptions = createChildrenOptions<App>({
+  useActions: () => [
+    {
+      key: "open",
+      component: () => (
+        <Group mx="md" my="sm">
+          <IconExternalLink stroke={1.5} />
+          <Text>Open app url</Text>
```
Comment (Meierschlumpf) [inline packages/spotlight/src/modes/app-integration-board/apps-search-group.tsx:21]: Translate all action components

File: packages/spotlight/src/modes/app-integration-board/apps-search-group.tsx [hunk 13]
```diff
@@ -0,0 +1,80 @@
...
+          <IconEye stroke={1.5} />
+          <Text>Edit app</Text>
+        </Group>
+      ),
+      useInteraction: interaction.link(({ id }) => ({ href: `/manage/apps/edit/${id}` })),
+    },
+  ],
+  detailComponent: ({ options }) => (
+    <Stack mx="md" my="sm">
+      <Text>Select an action for the app</Text>
```
Comment (Meierschlumpf) [inline packages/spotlight/src/modes/app-integration-board/apps-search-group.tsx:39]: Translate all detail components

File: packages/spotlight/src/modes/app-integration-board/apps-search-group.tsx [hunk 14]
```diff
@@ -0,0 +1,80 @@
...
+      <Avatar
+        size="sm"
+        src={app.iconUrl}
+        radius={0}
+        styles={{
+          image: {
+            objectFit: "contain",
+          },
+        }}
+      />{" "}
```
Comment (Meierschlumpf) [inline packages/spotlight/src/modes/app-integration-board/apps-search-group.tsx:72]: Probably not wanted `{" "}`

File: packages/spotlight/src/modes/app-integration-board/index.tsx [hunk 15]
```diff
@@ -0,0 +1,18 @@
...
+import { boardsSearchGroup } from "./boards-search-group";
+import { integrationsSearchGroup } from "./integrations-search-group";
+
+export const appIntegrationBoardMode = {
+  name: "appIntegrationBoard",
+  character: "#",
+  help: "Search for apps, integrations and boards",
+  tip: (
+    <Text size="xs" c="gray.6">
+      TODO: Add tip
```
Comment (Meierschlumpf) [inline packages/spotlight/src/modes/app-integration-board/index.tsx:14]: Add tips where not yet added

File: packages/spotlight/src/lib/mode.ts (hunk 11 above)
Comment (Meierschlumpf) [reply packages/spotlight/src/lib/mode.ts:6]: Unnecessary as it's only a key, renamed to `key`

File: packages/spotlight/src/components/actions/items/group-action-item.tsx [hunk 16]
```diff
@@ -0,0 +1,53 @@
...
+      : undefined;
+
+  const onClick =
+    interaction.type === "javaScript"
+      ? interaction.onSelect
+      : interaction.type === "mode"
+        ? () => setMode(interaction.mode)
+        : interaction.type === "children"
+          ? () => setChildrenOptions(interaction)
+          : undefined;
```
Comment (manuel-rw) [inline packages/spotlight/src/components/actions/items/group-action-item.tsx:41]: Can you simplify? This is hard to read

File: packages/spotlight/src/components/spotlight.tsx [hunk 17]
```diff
@@ -0,0 +1,117 @@
...
+      onQueryChange={(query) => {
+        if (mode === "help" && query.length === 1) {
+          const modeToActivate = searchModes.find((mode) => mode.character === query);
+          if (modeToActivate) {
+            setMode(modeToActivate.modeKey);
+            setQuery("");
+            setTimeout(() => selectAction(0, spotlightStore));
+            return;
+          }
+        }
```
Comment (manuel-rw) [inline packages/spotlight/src/components/spotlight.tsx:45]: Invert to reduce nesting?

**Review by manuel-rw — CHANGES_REQUESTED** [review]
This is so cool! Thank you. Please also have a look at DeepSource

**Comment by Meierschlumpf** [comment]
> This is so cool! Thank you. Please also have a look at DeepSource

Thank you, I also really like it, in a next step I'll add a way to customize the search engines and add a default one. Then I'll add a start page where you can search with that engine or whatever is configured for the specific page (Probably will add a useRegisterSpotlightAction hook or something. So for example the apps of a board are shown directly on the start page

**Review by manuel-rw — APPROVED** [review]

**Pull request merged**